File: python/Find the Duplicate Number.py
class Solution(object):
    def findDuplicate_binary(self, nums):
        left, right, count = 1, len(nums) - 1, 0
        while left < right:
            mid = (left + right) / 2
            count = 0
            for item in nums:
                if item <= mid:
                    count += 1
            
            if count > mid:
                right = mid
            else:
                left = mid + 1
        
        return (left + right) / 2
        
    def findDuplicate(self, nums):
        if len(nums) > 1:
            slow, fast = nums[0], nums[nums[0]]
            while slow != fast:
                slow = nums[slow]
                fast = nums[nums[fast]]
            
            fast = 0
            
            while slow != fast:
                slow = nums[slow]
                fast = nums[fast]
            
            return slow

File: cpp/Find the Duplicate Number.cpp
class Solution {
public:
    int findDuplicate_cheat(vector<int>& nums) {
        sort(nums.begin(), nums.end());
        for (int i = 0; i < nums.size() - 1; i++) {
            if (nums[i] == nums[i + 1]) {
                return nums[i];
            }
        }
    }
    
    int findDuplicate(vector<int>& nums) {
        int low = 1, high = nums.size() - 1;
        while (low < high) {
            int mid = (low + high) / 2;
            int count = 0;
            for (int i = 0; i < nums.size(); i++) {
                if (nums[i] <= mid) {
                    count++;
                }
            }
            if (count > mid) {
                high = mid;
            } else {
                low = mid + 1;
            }
        }
        return (low + high) / 2;
    }
};
